src/tiff/man/tiffset.1
.\" $Id: tiffset.1,v 1.7 2016-09-25 20:05:52 bfriesen Exp $
.\"
.\" Copyright (c) 1988-1997 Sam Leffler
.\" Copyright (c) 1991-1997 Silicon Graphics, Inc.
.\"
.\" Permission to use, copy, modify, distribute, and sell this software and 
.\" its documentation for any purpose is hereby granted without fee, provided
.\" that (i) the above copyright notices and this permission notice appear in
.\" all copies of the software and related documentation, and (ii) the names of
.\" Sam Leffler and Silicon Graphics may not be used in any advertising or
.\" publicity relating to the software without the specific, prior written
.\" permission of Sam Leffler and Silicon Graphics.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND, 
.\" EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY 
.\" WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  
.\" 
.\" IN NO EVENT SHALL SAM LEFFLER OR SILICON GRAPHICS BE LIABLE FOR
.\" ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
.\" OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
.\" WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF 
.\" LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE 
.\" OF THIS SOFTWARE.
.\"
.if n .po 0
.TH TIFFSET 1 "December 3, 2012" "libtiff"
.SH NAME
tiffset \- set or unset a field in a
.SM TIFF
header
.SH SYNOPSIS
.B tiffset
[
.I options
]
.I filename.tif
.SH DESCRIPTION
.I Tiffset
sets the value of a
.SM TIFF
header to a specified value or removes an existing setting.
.SH OPTIONS
.TP
.BI \-d " dirnumber"
change the current directory (starting at 0).
.TP
.BI \-s " tagnumber" "\fR [\fP" " count" "\fR ]\fP" " value ..."
Set the value of the named tag to the value or values specified.
.TP
.BI \-sd " diroffset"
change the current directory by offset.
.TP
.BI \-sf " tagnumber filename"
Set the value of the tag to the contents of filename.  This option is
supported for ASCII tags only.
.TP
.BI \-u " tagnumber"
Unset the tag.
.SH EXAMPLES
The following example sets the image description tag (270) of a.tif to
the contents of the file descrip:
.RS
.nf
tiffset \-sf 270 descrip a.tif
.fi
.RE
.PP
The following example sets the artist tag (315) of a.tif to the string
``Anonymous'':
.RS
.nf
tiffset \-s 315 Anonymous a.tif
.fi
.RE
.PP
This example sets the resolution of the file a.tif to 300 dpi:
.RS
.nf
tiffset \-s 296 2 a.tif
tiffset \-s 282 300.0 a.tif
tiffset \-s 283 300.0 a.tif
.fi
.RE
.PP
Set the photometric interpretation of the third page of a.tif to
min-is-black (ie. inverts it):
.RS
.nf
tiffset -d 2 -s 262 1 a.tif
.fi
.RE
.SH "SEE ALSO"
.BR tiffdump (1),
.BR tiffinfo (1),
.BR tiffcp (1),
.BR libtiff (3TIFF)
.PP
Libtiff library home page:
.BR http://www.simplesystems.org/libtiff/


debian/changelog
wxwidgets2.8 (2.8.1.1-0) unstable; urgency=low

  * Fix compilation with wxUSE_STL=1
  * wxGrid::GetBestSize() returns same size the grid would have after AutoSize()
  * Added wxTreeCtrl::CollapseAll[Children]() and IsEmpty() (Francesco Montorsi)
  * Several RTL-related positioning fixes (Diaa Sami)
  * Fix wxConfig::DeleteGroup() for arguments with trailing slash (David Hart)
  * Fix memory leak in wxGrid::ShowCellEditControl() (Christian Sturmlechner)
  * Don't crash if command line is not valid UTF-8 (Unicode build only)
  * Added wxSizerFlags::Shaped() and FixedMinSize() methods
  * Added wxCSConv::IsOk() (Manuel Martin)
  * Added wxDateTime::GetDateOnly()
  * Made wxTextFile work with unseekable files again (David Hart)
  * Added wxCONFIG_USE_SUBDIR flag to wxFileConfig (Giuseppe Bilotta)
  * Added wxSearchCtrl::[Get|Set]DescriptiveText

 -- Robin Dunn <robin@alldunn.com>  Fri, 19 Jan 2007 12:02:52 -0800

wxwidgets2.8 (2.8.0.1-1) unstable; urgency=low

  * No changes, just a different packager

 -- Vadim Zeitlin <vadim@wxwindows.org>  Dec, 13 Wed 2006 17:40:24 +0100

wxwidgets2.8 (2.8.0.1-0) unstable; urgency=low

  * Add wxSearchCtrl
  * Official wxPython release

 -- Robin Dunn <robin@alldunn.com>  Mon, 11 Dec 2006 18:50:29 -0800

wxwidgets2.8 (2.8.0.0-0rc3) unstable; urgency=low

  * wxFileSystem::OpenFile() takes wxFS_SEEKABLE by default now.
  * Fixed wxCalendarCtrl::HitTest() when clicking on month change arrows.
  * Added wxWindow::GetWindowBorderSize(), corrected wxTreeCtrl::GetBestSize().
  * Allow dynamically changing most of text control styles.
  * Enable use of libgnomeprintui by default in configure.

 -- Vadim Zeitlin <vadim@wxwindows.org>  Dec, 05 Tue 2006 14:41:22 +0100

wxwidgets2.8 (2.8.0.0-0rc2) unstable; urgency=low

  * wxArrayString::Alloc() doesn't clear array contents any more
  * Fixed long standing bug in wxFileConfig groups renaming

 -- Vadim Zeitlin <vadim@wxwindows.org>  Nov, 14 Tue 2006 19:06:37 +0100

wxwidgets2.7 (2.7.2.0-0) unstable; urgency=low

  * Added wxTar streams.
  * Added wxFilterFSHandler and wxArchiveFSHandler.
  * Added wxString::ToLongLong() and ToULongLong().
  * Added wxCollapsiblePane.
  * Added wxSimpleHtmlListBox.
  * Many enhancements to wxRichTextCtrl.
  * Support for loading TGA files added.
  * wxHtmlWindow now generates events for mouse actions and text copy.

 -- Vadim Zeitlin <vadim@wxwindows.org>  Oct, 31 Tue 2006 15:58:00 +0100

wxwidgets2.7 (2.7.1.3-0) unstable; urgency=low

  * wxGTK:  The wx.ALWAYS_SHOW_SB style is now supported.
  * Fixed name errors in the old wxPython package namespace.

 -- Robin Dunn <robin@alldunn.com>  Thu, 26 Oct 2006 14:35:29 -0700

wxwidgets2.7 (2.7.1.2-0) unstable; urgency=low

  * Fixed a bug in the MaskedEdit controls caused by conflicting IsEmpty methods.
  * Patch #1579280: Some mimetype optimizations on unix-like systems.

 -- Robin Dunn <robin@alldunn.com>  Sat, 21 Oct 2006 22:33:41 -0700

wxwidgets2.7 (2.7.1.1-0) unstable; urgency=low

  * First 2.7.x release from wxPython team.
  * Many enhancements, changes, fixes and updates.  See
    http://wxpython.org/CHANGES.txt for details.

 -- Robin Dunn <robin@alldunn.com>  Wed, 18 Oct 2006 17:09:53 -0700

wxwidgets2.7 (2.7.0.0-2) unstable; urgency=low

  * Fixes to ansi packages builds
  * Corresponds to the official wxWidgets 2.7.0 sources

 -- Vadim Zeitlin <vadim@wxwindows.org>  Fri,  6 Aug 2006 16:31:08 +0200

wxwidgets2.7 (2.7.0.0-1) unstable; urgency=low

  * New 2.7.0 release of the new development series.
  * It's now possible to build both ansi and unicode versions.
  * Maintainer change.

 -- Vadim Zeitlin <vadim@wxwindows.org>  Fri,  4 Aug 2006 16:12:47 +0200

wxwidgets2.7 (2.7.0.0-0rpd) unstable; urgency=low

  * First test build with wx 2.7 tree.

 -- Robin Dunn <robin@alldunn.com>  Thu,  1 Jun 2006 14:23:16 -0700

wxwidgets2.6 (2.6.3.2-3rpd) unstable; urgency=low

  * Added icons and *.desktop files for the wpython-wxtools

 -- Robin Dunn <robin@alldunn.com>  Thu,  4 May 2006 15:35:08 -0700

wxwidgets2.6 (2.6.3.2) unstable; urgency=low

  * Next try for an upload candidate.
  * Fixes FTBFS on GNU/kFreeBSD. Closes: #327638
  * Update fr.po, thanks to Thomas Huriaux and debian-l10n-french.
    Closes: #328966
  * Split wxbase back into its own package, after all these years
    and just after I drop it, someone finally has a use for it.
    Such is life.  Closes: #333944
  * Fixes FTBFS with new make.  If this ugly quoting is the tradeoff,
    I'm itching to learn of the fabulous benefits...
    Ack NMU for this from Kurt Roeckx to keep the amd64 transition
    rolling while my x86 build machine was painfully dying.  Thanks!
    Closes: #350695, #361443
  * New FSF address in copyright file.
  * Add an index.html link to the doc contents page.  Closes: #328896
  * Fixes declarations for friends, as required by C++, and now gcc4.1.
    Closes: #358219
  * Add an unpack_examples script, loosely based on one submitted
    by Richard Esplin.  Closes: #342761
  * Disable the reserved vtable entries (again, but this time its
    optional) to retain compatibility with the existing release
    in the distro.
  * Update build-deps for xorg7.

 -- Ron Lee <ron@debian.org>  Fri, 14 Apr 2006 16:50:24 +0930

wxwidgets2.6 (2.6.1.2-0.1) unstable; urgency=low

  * Non-maintainer upload.
  * Make it build using the new make.  Patch from
    Daniel Schepler <schepler@math.berkeley.edu> (Closes: #350695)

 -- Kurt Roeckx <kurt@roeckx.be>  Sun,  9 Apr 2006 14:04:34 +0200

wxwidgets2.6 (2.6.1.2) unstable; urgency=low

  * Fix the build_options.py created by the in tree build.
    Closes: #324960  Thanks.
  * Move the wx.pth alternative to its own directory where
    it can't mess with anything else.
  * Pull in changes to dcclient.cpp and window.cpp from HEAD
    mostly for gtk2.8 compatibility, but fixes a couple of
    other issues too.

 -- Ron Lee <ron@debian.org>  Thu, 25 Aug 2005 18:38:31 +0930

wxwidgets2.6 (2.6.1.1.1) unstable; urgency=low

src/tiff/tools/tiffset.c
/******************************************************************************
 * $Id: tiffset.c,v 1.18 2012-12-04 03:02:37 bfriesen Exp $
 *
 * Project:  libtiff tools
 * Purpose:  Mainline for setting metadata in existing TIFF files.
 * Author:   Frank Warmerdam, warmerdam@pobox.com
 *
 ******************************************************************************
 * Copyright (c) 2000, Frank Warmerdam
 *
 * Permission to use, copy, modify, distribute, and sell this software and 
 * its documentation for any purpose is hereby granted without fee, provided
 * that (i) the above copyright notices and this permission notice appear in
 * all copies of the software and related documentation, and (ii) the names of
 * Sam Leffler and Silicon Graphics may not be used in any advertising or
 * publicity relating to the software without the specific, prior written
 * permission of Sam Leffler and Silicon Graphics.
 * 
 * THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND, 
 * EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY 
 * WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  
 * 
 * IN NO EVENT SHALL SAM LEFFLER OR SILICON GRAPHICS BE LIABLE FOR
 * ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
 * OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
 * WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF 
 * LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE 
 * OF THIS SOFTWARE.
 ******************************************************************************
 */


#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "tiffio.h"

static char* usageMsg[] = {
"usage: tiffset [options] filename",
"where options are:",
" -s <tagname> [count] <value>...   set the tag value",
" -u <tagname> to unset the tag",
" -d <dirno> set the directory",
" -sd <diroff> set the subdirectory",
" -sf <tagname> <filename>  read the tag value from file (for ASCII tags only)",
NULL
};

static void
usage(void)
{
	int i;
	for (i = 0; usageMsg[i]; i++)
		fprintf(stderr, "%s\n", usageMsg[i]);
	exit(-1);
}

static const TIFFField *
GetField(TIFF *tiff, const char *tagname)
{
    const TIFFField *fip;

    if( atoi(tagname) > 0 )
        fip = TIFFFieldWithTag(tiff, (ttag_t)atoi(tagname));
    else
        fip = TIFFFieldWithName(tiff, tagname);

    if (!fip) {
        fprintf( stderr, "Field name \"%s\" is not recognised.\n", tagname );
        return (TIFFField *)NULL;
    }

    return fip;
}

int
main(int argc, char* argv[])
{
    TIFF *tiff;
    int  arg_index;

    if (argc < 2)
        usage();

    tiff = TIFFOpen(argv[argc-1], "r+");
    if (tiff == NULL)
        return 2;

    for( arg_index = 1; arg_index < argc-1; arg_index++ ) {
	if (strcmp(argv[arg_index],"-d") == 0 && arg_index < argc-2) {
	    arg_index++;
	    if( TIFFSetDirectory(tiff, atoi(argv[arg_index]) ) != 1 )
            {
               fprintf( stderr, "Failed to set directory=%s\n", argv[arg_index] );
               return 6;
            }
	    arg_index++;
	}
	if (strcmp(argv[arg_index],"-sd") == 0 && arg_index < argc-2) {
	    arg_index++;
	    if( TIFFSetSubDirectory(tiff, atoi(argv[arg_index]) ) != 1 )
            {
               fprintf( stderr, "Failed to set sub directory=%s\n", argv[arg_index] );
               return 7;
            }
	    arg_index++;
	}
    /* Add unset option to tiffset -- Zach Baker (niquil@niquil.net) 11/14/2012 */ 
    if (strcmp(argv[arg_index],"-u") == 0 && arg_index < argc-2) {
            const TIFFField *fip;
            const char *tagname;
            arg_index++;
            tagname = argv[arg_index];
            fip = GetField(tiff, tagname);
            if (!fip)
                return 3;

            if (TIFFUnsetField(tiff, TIFFFieldTag(fip)) != 1)
            {
                    fprintf(stderr, "Failed to unset %s\n", TIFFFieldName(fip));
            }
            arg_index++;
    } else if (strcmp(argv[arg_index],"-s") == 0 && arg_index < argc-3) {
            const TIFFField *fip;
            const char *tagname;

            arg_index++;
            tagname = argv[arg_index];
            fip = GetField(tiff, tagname);

            if (!fip)
                return 3;

            arg_index++;
            if (TIFFFieldDataType(fip) == TIFF_ASCII) {
                if (TIFFSetField(tiff, TIFFFieldTag(fip), argv[arg_index]) != 1)
                    fprintf( stderr, "Failed to set %s=%s\n",
                             TIFFFieldName(fip), argv[arg_index] );
            } else if (TIFFFieldWriteCount(fip) > 0
		       || TIFFFieldWriteCount(fip) == TIFF_VARIABLE) {
                int     ret = 1;
                short   wc;

                if (TIFFFieldWriteCount(fip) == TIFF_VARIABLE)
                        wc = atoi(argv[arg_index++]);
                else
                        wc = TIFFFieldWriteCount(fip);

                if (argc - arg_index < wc) {
                    fprintf( stderr,
                             "Number of tag values is not enough. "
                             "Expected %d values for %s tag, got %d\n",
                             wc, TIFFFieldName(fip), argc - arg_index);
                    return 4;
                }
                    
                if (wc > 1) {
                        int     i, size;
                        void    *array;

                        switch (TIFFFieldDataType(fip)) {
                                /*
                                 * XXX: We can't use TIFFDataWidth()
                                 * to determine the space needed to store
                                 * the value. For TIFF_RATIONAL values
                                 * TIFFDataWidth() returns 8, but we use 4-byte
                                 * float to represent rationals.
                                 */
                                case TIFF_BYTE:
                                case TIFF_ASCII:
                                case TIFF_SBYTE:
                                case TIFF_UNDEFINED:
				default:
                                    size = 1;
                                    break;

                                case TIFF_SHORT:
                                case TIFF_SSHORT:
                                    size = 2;
                                    break;

                                case TIFF_LONG:
                                case TIFF_SLONG:
                                case TIFF_FLOAT:
                                case TIFF_IFD:
                                case TIFF_RATIONAL:
                                case TIFF_SRATIONAL:
                                    size = 4;
                                    break;

                                case TIFF_DOUBLE:
                                    size = 8;
                                    break;
                        }

                        array = _TIFFmalloc(wc * size);
                        if (!array) {
                                fprintf(stderr, "No space for %s tag\n",
                                        tagname);
                                return 4;
                        }

                        switch (TIFFFieldDataType(fip)) {
                            case TIFF_BYTE:
                                for (i = 0; i < wc; i++)
                                    ((uint8 *)array)[i] = atoi(argv[arg_index+i]);
                                break;
                            case TIFF_SHORT:
                                for (i = 0; i < wc; i++)
                                    ((uint16 *)array)[i] = atoi(argv[arg_index+i]);
                                break;
                            case TIFF_SBYTE:
                                for (i = 0; i < wc; i++)
                                    ((int8 *)array)[i] = atoi(argv[arg_index+i]);
                                break;
                            case TIFF_SSHORT:
                                for (i = 0; i < wc; i++)
                                    ((int16 *)array)[i] = atoi(argv[arg_index+i]);
                                break;
                            case TIFF_LONG:
                                for (i = 0; i < wc; i++)
                                    ((uint32 *)array)[i] = atol(argv[arg_index+i]);
                                break;
                            case TIFF_SLONG:
                            case TIFF_IFD:
                                for (i = 0; i < wc; i++)
                                    ((uint32 *)array)[i] = atol(argv[arg_index+i]);
                                break;
                            case TIFF_DOUBLE:
                                for (i = 0; i < wc; i++)
                                    ((double *)array)[i] = atof(argv[arg_index+i]);
                                break;
                            case TIFF_RATIONAL:
                            case TIFF_SRATIONAL:
                            case TIFF_FLOAT:
                                for (i = 0; i < wc; i++)
                                    ((float *)array)[i] = (float)atof(argv[arg_index+i]);
                                break;
                            default:
                                break;
                        }
                
                        if (TIFFFieldPassCount(fip)) {
                                ret = TIFFSetField(tiff, TIFFFieldTag(fip),
                                                   wc, array);
                        } else if (TIFFFieldTag(fip) == TIFFTAG_PAGENUMBER
				   || TIFFFieldTag(fip) == TIFFTAG_HALFTONEHINTS
				   || TIFFFieldTag(fip) == TIFFTAG_YCBCRSUBSAMPLING
				   || TIFFFieldTag(fip) == TIFFTAG_DOTRANGE) {
       				if (TIFFFieldDataType(fip) == TIFF_BYTE) {
					ret = TIFFSetField(tiff, TIFFFieldTag(fip),
						((uint8 *)array)[0], ((uint8 *)array)[1]);
				} else if (TIFFFieldDataType(fip) == TIFF_SHORT) {
					ret = TIFFSetField(tiff, TIFFFieldTag(fip),
						((uint16 *)array)[0], ((uint16 *)array)[1]);
				}
			} else {
                                ret = TIFFSetField(tiff, TIFFFieldTag(fip),
                                                   array);
                        }

                        _TIFFfree(array);
                } else {
                        switch (TIFFFieldDataType(fip)) {
                            case TIFF_BYTE:
                            case TIFF_SHORT:
                            case TIFF_SBYTE:
                            case TIFF_SSHORT:
                                ret = TIFFSetField(tiff, TIFFFieldTag(fip),
                                                   atoi(argv[arg_index++]));
                                break;
                            case TIFF_LONG:
                            case TIFF_SLONG:
                            case TIFF_IFD:
                                ret = TIFFSetField(tiff, TIFFFieldTag(fip),
                                                   atol(argv[arg_index++]));
                                break;
                            case TIFF_DOUBLE:
                                ret = TIFFSetField(tiff, TIFFFieldTag(fip),
                                                   atof(argv[arg_index++]));
                                break;
                            case TIFF_RATIONAL:
                            case TIFF_SRATIONAL:
                            case TIFF_FLOAT:
                                ret = TIFFSetField(tiff, TIFFFieldTag(fip),
                                                   (float)atof(argv[arg_index++]));
                                break;
                            default:
                                break;
                        }
                }

                if (ret != 1)
                    fprintf(stderr, "Failed to set %s\n", TIFFFieldName(fip));
                arg_index += wc;
            }
        } else if (strcmp(argv[arg_index],"-sf") == 0 && arg_index < argc-3) {
            FILE    *fp;
            const TIFFField *fip;
            char    *text;
            size_t  len;

            arg_index++;
            fip = GetField(tiff, argv[arg_index]);

            if (!fip)
                return 3;

            if (TIFFFieldDataType(fip) != TIFF_ASCII) {
                fprintf( stderr,
                         "Only ASCII tags can be set from file. "
                         "%s is not ASCII tag.\n", TIFFFieldName(fip) );
                return 5;
            }

            arg_index++;
            fp = fopen( argv[arg_index], "rt" );
            if(fp == NULL) {
                perror( argv[arg_index] );
                continue;
            }

            text = (char *) malloc(1000000);
            len = fread( text, 1, 999999, fp );
            text[len] = '\0';

            fclose( fp );

            if(TIFFSetField( tiff, TIFFFieldTag(fip), text ) != 1) {
                fprintf(stderr, "Failed to set %s from file %s\n", 
                        TIFFFieldName(fip), argv[arg_index]);
            }

            _TIFFfree( text );
            arg_index++;
        } else {
            fprintf(stderr, "Unrecognised option: %s\n",
                    argv[arg_index]);
            usage();
        }
    }

    TIFFRewriteDirectory(tiff);
    TIFFClose(tiff);
    return 0;
}

/* vim: set ts=8 sts=8 sw=8 noet: */
/*
 * Local Variables:
 * mode: c
 * c-basic-offset: 8
 * fill-column: 78
 * End:
 */
